core/src/packet_forward_manager.rs

```diff
@@ -0,0 +1,427 @@
+use {
+    crate::{
+        banking_stage::TOTAL_BUFFERED_PACKETS, qos_service::QosService,
+        unprocessed_packet_batches::*,
+    },
+    solana_perf::packet::Packet,
+    solana_runtime::{bank::Bank, cost_model::TransactionCost, cost_tracker::CostTracker},
+    solana_sdk::transaction::SanitizedTransaction,
+    std::sync::Arc,
+};
+
+#[derive(Debug)]
+pub struct PacketForwardManager<'a> {
+    cost_trackers: Vec<CostTracker>,
+    forwardable_packets: Vec<Vec<&'a Packet>>,
+}
+
+impl<'a> PacketForwardManager<'a> {
+    pub fn new(max_forward_block_count: u64) -> Self {
+        Self {
+            cost_trackers: (0..max_forward_block_count)
+                .into_iter()
+                .map(|_| CostTracker::default())
+                .collect(),
+            forwardable_packets: (0..max_forward_block_count)
+                .into_iter()
+                .map(|_| Vec::<&'a Packet>::new())
+                .collect(),
+        }
+    }
+
+    pub fn take(
+        &mut self,
+        unprocessed_packet_batches: &'a UnprocessedPacketBatches,
+        working_bank: Arc<Bank>,
+        qos_service: &'a QosService,
+    ) -> Vec<&'a Packet> {
+        let prioritized_forwardable_packet_locators = Self::prioritize_unforwarded_packets_by_fee(
+            unprocessed_packet_batches,
+            Some(working_bank.clone()),
+        );
+
+        // if we have a bank to work with, then sort packets by write-account buckets
+        let (transactions, sanitized_locators) = sanitize_transactions(
+            unprocessed_packet_batches,
+            &prioritized_forwardable_packet_locators,
+            &working_bank.feature_set,
+            working_bank.vote_only_bank(),
+            working_bank.as_ref(),
+        );
+        let transactions_costs = qos_service.compute_transaction_costs(transactions.iter());
```

Comment:
Let's add a comment describing the steps here:

1. Compute the fee per cu and sort transactions by this priority for the `unprocessed_packet_batches` queue in `prioritize_unforwarded_packets_by_fee()`
2. Lock cost model, compute cost model cost in `compute_transaction_costs()`
3. Using the cost model costs from 2. split the `unprocessed_packet_batches` into separate chunks, where each chunk represents a block filled with transactions with costs that add up to the block cost limit

Reply:
Some perf implications: 

1. If the `unprocessed_packet_batches` was already a heap sorted by fee-per-cu priority which was updated when new transactions were added, we could avoid the sort here
2. Unprocessed buffer could be huge, calculating all this metadata might be expensive. Because there's already a limit on the number of packet bytes we actually forward here: https://github.com/solana-labs/solana/blob/86e2f728c3ef6ff18d605d4f0e15c02ba295a286/core/src/banking_stage.rs#L483-L500, it might be good at the beginning to truncate the buffer up to a sane number of packets that will actually be forwarded